File: frontend/src/test/catalog.ts
import type { CatalogDTO } from "@/queries/catalog";

/** TEST_CATALOG mirrors the Wails catalog DTO with a distinctive subset of
 * production values. Page tests assert that forms render exactly these
 * options so a leftover hardcoded array cannot hide behind the mock. */
export const TEST_CATALOG: CatalogDTO = {
  currencies: ["USD", "SGD", "CNY"],
  instrumentTypes: ["stock", "etf"],
  institutionTypes: ["bank", "brokerage", "insurer", "exchange", "employer", "government", "other"],
  quoteSources: ["manual", "provider"],
  instrumentProviders: ["yahoo_finance"],
  accountTypes: ["cash_on_hand", "bank_account", "brokerage", "credit_card", "other"],
  balanceSheetRoles: ["asset", "liability"],
  trackingModes: ["balance", "manual_value", "holdings"],
  accountCombinations: [
    { accountType: "cash_on_hand", balanceSheetRole: "asset", trackingMode: "balance", roleLocked: true, includeInNetWorth: true, includeInPortfolio: false, includeInLiquidAssets: true, wholeAccountWarning: false },
    { accountType: "bank_account", balanceSheetRole: "asset", trackingMode: "balance", roleLocked: true, includeInNetWorth: true, includeInPortfolio: false, includeInLiquidAssets: true, wholeAccountWarning: false },
    { accountType: "bank_account", balanceSheetRole: "asset", trackingMode: "holdings", roleLocked: true, includeInNetWorth: true, includeInPortfolio: false, includeInLiquidAssets: false, wholeAccountWarning: true },
    { accountType: "brokerage", balanceSheetRole: "asset", trackingMode: "holdings", roleLocked: true, includeInNetWorth: true, includeInPortfolio: true, includeInLiquidAssets: false, wholeAccountWarning: true },
    { accountType: "brokerage", balanceSheetRole: "asset", trackingMode: "manual_value", roleLocked: true, includeInNetWorth: true, includeInPortfolio: true, includeInLiquidAssets: false, wholeAccountWarning: false },
    { accountType: "credit_card", balanceSheetRole: "liability", trackingMode: "balance", roleLocked: true, includeInNetWorth: true, includeInPortfolio: false, includeInLiquidAssets: false, wholeAccountWarning: false },
    { accountType: "other", balanceSheetRole: "asset", trackingMode: "balance", roleLocked: false, includeInNetWorth: true, includeInPortfolio: false, includeInLiquidAssets: false, wholeAccountWarning: false },
    { accountType: "other", balanceSheetRole: "liability", trackingMode: "balance", roleLocked: false, includeInNetWorth: true, includeInPortfolio: false, includeInLiquidAssets: false, wholeAccountWarning: false },
  ],
  trackingModesByAccountType: {
    cash_on_hand: ["balance"],
    bank_account: ["balance", "holdings"],
    brokerage: ["holdings", "manual_value"],
    credit_card: ["balance"],
    other: ["balance"],
  },
  trendRanges: ["30d", "1y"],
  appearances: ["system", "light", "dark"],
  languages: ["system", "en", "zh-CN", "zh-TW"],
  accents: ["nestworth", "ocean", "forest", "amber", "rose"],
  moneyInReasons: ["income", "contribution", "gift", "other", "reconciliation"],
  moneyOutReasons: ["expense", "fee", "tax", "other", "reconciliation"],
  valueUpdateReasons: ["reconciliation", "other"],
  tradeSides: ["buy", "sell"],
};

export function selectValues(element: HTMLElement): string[] {
  return Array.from(element.querySelectorAll("option")).map((option) => (option as HTMLOptionElement).value);
}

File: internal/wailsapi/catalog/catalog.go
package catalog

import (
	"github.com/waltwang/nestworth-go/internal/application"
	"github.com/waltwang/nestworth-go/internal/domain"
	"github.com/waltwang/nestworth-go/internal/settings"
)

type Service struct{}

func NewService() *Service {
	return &Service{}
}

type AccountCombinationDTO struct {
	AccountType           string `json:"accountType"`
	BalanceSheetRole      string `json:"balanceSheetRole"`
	TrackingMode          string `json:"trackingMode"`
	RoleLocked            bool   `json:"roleLocked"`
	IncludeInNetWorth     bool   `json:"includeInNetWorth"`
	IncludeInPortfolio    bool   `json:"includeInPortfolio"`
	IncludeInLiquidAssets bool   `json:"includeInLiquidAssets"`
	WholeAccountWarning   bool   `json:"wholeAccountWarning"`
}

type CatalogDTO struct {
	Currencies                 []string                `json:"currencies"`
	InstrumentTypes            []string                `json:"instrumentTypes"`
	QuoteSources               []string                `json:"quoteSources"`
	InstrumentProviders        []string                `json:"instrumentProviders"`
	AccountTypes               []string                `json:"accountTypes"`
	BalanceSheetRoles          []string                `json:"balanceSheetRoles"`
	TrackingModes              []string                `json:"trackingModes"`
	AccountCombinations        []AccountCombinationDTO `json:"accountCombinations"`
	TrackingModesByAccountType map[string][]string     `json:"trackingModesByAccountType"`
	TrendRanges                []string                `json:"trendRanges"`
	Appearances                []string                `json:"appearances"`
	Languages                  []string                `json:"languages"`
	Accents                    []string                `json:"accents"`
	MoneyInReasons             []string                `json:"moneyInReasons"`
	MoneyOutReasons            []string                `json:"moneyOutReasons"`
	ValueUpdateReasons         []string                `json:"valueUpdateReasons"`
	TradeSides                 []string                `json:"tradeSides"`
}

func (s *Service) Catalog() CatalogDTO {
	combinations := make([]AccountCombinationDTO, 0, len(domain.LegalAccountCombinations()))
	trackingByType := map[string][]string{}
	seenTracking := map[string]map[string]struct{}{}
	for _, combination := range domain.LegalAccountCombinations() {
		defaults := domain.SuggestedInclusion(combination.AccountType, combination.TrackingMode)
		_, roleLocked := domain.DefaultBalanceSheetRole(combination.AccountType)
		item := AccountCombinationDTO{
			AccountType:           combination.AccountType.String(),
			BalanceSheetRole:      combination.BalanceSheetRole.String(),
			TrackingMode:          string(combination.TrackingMode),
			RoleLocked:            roleLocked,
			IncludeInNetWorth:     defaults.IncludeInNetWorth,
			IncludeInPortfolio:    defaults.IncludeInPortfolio,
			IncludeInLiquidAssets: defaults.IncludeInLiquidAssets,
			WholeAccountWarning:   combination.TrackingMode == domain.TrackingHoldings,
		}
		combinations = append(combinations, item)
		typeKey := combination.AccountType.String()
		if seenTracking[typeKey] == nil {
			seenTracking[typeKey] = map[string]struct{}{}
		}
		mode := string(combination.TrackingMode)
		if _, exists := seenTracking[typeKey][mode]; exists {
			continue
		}
		seenTracking[typeKey][mode] = struct{}{}
		trackingByType[typeKey] = append(trackingByType[typeKey], mode)
	}
	return CatalogDTO{
		Currencies:                 settings.SupportedCurrencies(),
		InstrumentTypes:            stringSlice(domain.AllInstrumentTypes()),
		QuoteSources:               stringSlice(domain.AllQuoteSourceKinds()),
		InstrumentProviders:        application.InstrumentProviderKeys(),
		AccountTypes:               stringSlice(domain.AllAccountTypes()),
		BalanceSheetRoles:          stringSlice(domain.AllBalanceSheetRoles()),
		TrackingModes:              stringSlice(domain.AllTrackingModes()),
		AccountCombinations:        combinations,
		TrackingModesByAccountType: trackingByType,
		TrendRanges:                stringSlice(domain.AllTrendRanges()),
		Appearances:                stringSlice(settings.AllAppearances()),
		Languages:                  stringSlice(settings.AllLanguages()),
		Accents:                    stringSlice(settings.AllAccents()),
		MoneyInReasons:             stringSlice(domain.MoneyInReasons()),
		MoneyOutReasons:            stringSlice(domain.MoneyOutReasons()),
		ValueUpdateReasons:         stringSlice(domain.ValueUpdateReasons()),
		TradeSides:                 stringSlice(domain.AllTradeSides()),
	}
}

func stringSlice[T ~string](values []T) []string {
	out := make([]string, len(values))
	for i, value := range values {
		out[i] = string(value)
	}
	return out
}
